roll/websocket.py

```diff
@@ -0,0 +1,124 @@
+    max_queue = 64
+    read_limit = 2 ** 16
+    write_limit = 2 ** 16
+
+    def __init__(self, handler, methods, subprotocols: list=None, **kwargs):
+        if methods != ['GET']:
+            raise RuntimeError('Websockets can only be registered on GET.')
+
+        self.handler = handler
+        if subprotocols:
+            subprotocols = frozenset(subprotocols)
+        self.subprotocols = subprotocols
+
+    def handshake(self, request):
+        """Websocket handshake, handled by `websockets`
+        """
+        headers = []
+
+        def get_header(k):
+            return request.headers.get(k.upper(), '')
+
+        def set_header(k, v):
+            headers.append((k, v))
+
+        try:
+            key = handshake.check_request(get_header)
+            handshake.build_response(set_header, key)
+        except InvalidHandshake:
+            raise RuntimeError('Invalid websocket request')
+
+        subprotocol = None
+        ws_protocol = get_header('Sec-Websocket-Protocol')
+        if self.subprotocols and ws_protocol:
+            # select a subprotocol
+            client_subprotocols = tuple(
+                (p.strip() for p in ws_protocol.split(',')))
+            for p in client_subprotocols:
+                if p in self.subprotocols:
+                    subprotocol = p
+                    set_header('Sec-Websocket-Protocol', subprotocol)
+                    break
+
+        # write the 101 response back to the client
+        # We will use a real Response in the future.
+        rv = b'HTTP/1.1 101 Switching Protocols\r\n'
+        for k, v in headers:
+            rv += k.encode('utf-8') + b': ' + v.encode('utf-8') + b'\r\n'
+        rv += b'\r\n'
+        request.transport.write(rv)
+
+        # Return the subprotocol agreed upon, if any
+        return subprotocol
+
+    def switch_protocol(self, request):
+        # The websocket handshake agrees on the subprotocol to use.
+        # It then writes a response to the client, before we go any further.
+        subprotocol = self.handshake(request)
+
+        # Creation of the new protocol, with the agreed upon subprotocol
+        websocket = WebSocketCommonProtocol(
+            timeout=self.timeout,
+            max_size=self.max_size,
+            max_queue=self.max_queue,
+            read_limit=self.read_limit,
+            write_limit=self.write_limit
+        )
+        websocket.subprotocol = subprotocol
+
+        # The protocol now opens the connection and gets pushed into place.
+        websocket.connection_made(request.transport)
+        websocket.connection_open()
+        request.transport.set_protocol(websocket)
+        return websocket
+
+    async def __call__(self, request, response, **params):
+        if request.upgrade != 'websocket':
+            # https://tools.ietf.org/html/rfc7231.html#page-62
+            # Upgrade needed but none was request or of the wrong type
+            response.status = HTTPStatus.UPGRADE_REQUIRED
```

Comment:
Why not raising a `roll.HttpError` here?

Reply:
For a sillyt cross import problem. WebsocketHandler is imported in the __init__.py file and thus importing from __init__ will create a problem, that's why I used the response, available there.

Reply:
Maybe we could give errors their dedicated file to avoid that? @yohanboniface thoughts?